File: packages/archkeep-rule-sdk-go/evidence_test.go
package archkeeprule

import (
	"errors"
	"strings"
	"testing"
)

// minimal is the smallest bundle that is still a whole one — every kind
// present, because the engine refuses to build a bundle with a kind missing.
const minimal = `{
	"contract": 1,
	"rule": { "name": "example-rule", "params": { "tag": "layer-domain" } },
	"model": { "projects": [{ "name": "alpha", "root": "packages/alpha", "tags": ["layer-domain"] }] },
	"graph": { "edges": [{ "source": "beta", "target": "alpha", "type": "static" }] },
	"imports": [{
		"sourceProject": "beta",
		"sourceFile": "packages/beta/src/lib.rs",
		"line": 3,
		"column": 5,
		"specifier": "alpha::thing",
		"kind": "static",
		"spelling": { "path": false, "relative": false },
		"resolved": { "target": "alpha", "file": null, "external": false, "packageName": null }
	}],
	"policy": { "depConstraints": [], "moduleBoundaryOptions": { "allow": [] } }
}`

func readMinimal(t *testing.T, mutations ...[2]string) *Evidence {
	t.Helper()
	document := minimal
	for _, mutation := range mutations {
		document = strings.Replace(document, mutation[0], mutation[1], 1)
	}
	evidence, err := ReadEvidence([]byte(document))
	if err != nil {
		t.Fatalf("the fixture is a bundle: %v", err)
	}
	return evidence
}

func TestAWholeBundleReadsEveryKind(t *testing.T) {
	evidence := readMinimal(t)

	if evidence.Rule.Name != "example-rule" {
		t.Fatalf("rule.name is %q", evidence.Rule.Name)
	}
	alpha, ok := evidence.Project("alpha")
	if !ok || !alpha.HasTag("layer-domain") {
		t.Fatal("alpha is declared and carries layer-domain")
	}
	if alpha.HasTag("layer-domai") {
		t.Fatal("a tag is compared exactly, never by prefix")
	}
	if _, ok := evidence.Project("beta"); ok {
		t.Fatal("beta is named by an edge and declared by no project")
	}
	if evidence.Graph.Edges[0].Type != "static" {
		t.Fatalf("edge type is %q", evidence.Graph.Edges[0].Type)
	}
	if evidence.Graph.Edges[0].SourceFile != nil {
		t.Fatal("an edge with no sourceFile carries nil, not the empty string")
	}
	if evidence.Imports[0].Kind != ImportStatic {
		t.Fatalf("import kind is %q", evidence.Imports[0].Kind)
	}
	if evidence.Imports[0].SourceProject != "beta" {
		t.Fatalf("attribution is %q", evidence.Imports[0].SourceProject)
	}
	resolved := evidence.Imports[0].Resolved
	if resolved == nil || resolved.Target == nil || *resolved.Target != "alpha" {
		t.Fatal("the import resolves into alpha")
	}
	if resolved.File != nil || resolved.PackageName != nil {
		t.Fatal("a null in the bundle is nil here, never the empty string")
	}
	if len(evidence.Policy.DepConstraints) != 0 {
		t.Fatal("the policy declares no constraint rows")
	}
	if string(evidence.Policy.ModuleBoundaryOptions) == "" {
		t.Fatal("the boundary options arrive as the raw document the engine loaded")
	}
}

func TestPositionsAreOneBasedAndZeroIsRefused(t *testing.T) {
	line, column, ok := readMinimal(t).Imports[0].Position()
	if !ok || line != 3 || column != 5 {
		t.Fatalf("the site is at 3:5, got %d:%d (ok=%v)", line, column, ok)
	}

	zeroed := readMinimal(t, [2]string{`"line": 3`, `"line": 0`})
	if _, _, ok := zeroed.Imports[0].Position(); ok {
		t.Fatal("a zeroth line is not a position, and clamping it would point at nothing")
	}
}

func TestTypedParamsReadTheRow(t *testing.T) {
	var declared struct {
		Tag string `json:"tag"`
	}
	if err := readMinimal(t).ReadParams(&declared); err != nil {
		t.Fatalf("the row declares a tag: %v", err)
	}
	if declared.Tag != "layer-domain" {
		t.Fatalf("params.tag is %q", declared.Tag)
	}
}

func TestAParamsFieldOfTheWrongTypeIsAnErrorRatherThanAnAbsentValue(t *testing.T) {
	evidence := readMinimal(t, [2]string{`"tag": "layer-domain"`, `"tag": 7`})
	var declared struct {
		Tag string `json:"tag"`
	}
	if err := evidence.ReadParams(&declared); err == nil {
		t.Fatal("a number where a tag belongs must not read as an absent tag")
	}
}

// TestABundleCarryingNoParamsIsNamedRatherThanParsedAsEmpty covers the one
// shape json.Unmarshal describes in the parser's words rather than the
// bundle's: the engine always writes {} for a row that declares none.
func TestABundleCarryingNoParamsIsNamedRatherThanParsedAsEmpty(t *testing.T) {
	evidence := readMinimal(t, [2]string{`"params": { "tag": "layer-domain" }`, `"params": null`})
	var declared struct{}
	err := evidence.ReadParams(&declared)
	if err == nil || !strings.Contains(err.Error(), "carries no rule.params") {
		t.Fatalf("an absent params must be named: %v", err)
	}
}

// The four below are the silent direction: none of these may read as a bundle,
// because every one of them would hand a rule facts it could then judge over
// and report clean.

func TestABundleFromAnotherContractIsRefusedByVersion(t *testing.T) {
	_, err := ReadEvidence([]byte(strings.Replace(minimal, `"contract": 1`, `"contract": 2`, 1)))
	var evidenceError *EvidenceError
	if !errors.As(err, &evidenceError) || evidenceError.Stated != 2 {
		t.Fatalf("a contract-2 bundle must be refused by version, got %v", err)
	}
	if !strings.Contains(err.Error(), "is not approximated") {
		t.Fatalf("the refusal names why: %v", err)
	}
}

func TestABundleMissingAKindIsRefused(t *testing.T) {
	// A kind cannot be "missing" in Go the way a missing serde field is: an
	// absent object decodes to the zero value. What the engine can never emit is
	// a kind whose SHAPE is wrong, and that is what a rule must not read past.
	_, err := ReadEvidence([]byte(strings.Replace(minimal, `"model": {`, `"model": [`, 1)))
	if err == nil {
		t.Fatal("a model that is not the model must not read as a bundle")
	}
}

func TestAnImportKindThisContractDoesNotCarryIsRefused(t *testing.T) {
	_, err := ReadEvidence([]byte(strings.Replace(minimal, `"kind": "static"`, `"kind": "ambient"`, 1)))
	if err == nil {
		t.Fatal("an import kind contract 1 does not carry must not read as one that it does")
	}
	if !strings.Contains(err.Error(), "ambient") {
		t.Fatalf("the refusal names the kind: %v", err)
	}
	var evidenceError *EvidenceError
	if !errors.As(err, &evidenceError) || evidenceError.Cause == nil {
		t.Fatalf("a malformed bundle names its cause: %v", err)
	}
}

func TestBytesThatAreNotABundleAreRefusedRatherThanZeroed(t *testing.T) {
	_, err := ReadEvidence([]byte("not json at all"))
	if err == nil || !strings.Contains(err.Error(), "could not be read as contract 1") {
		t.Fatalf("corrupt bytes must be refused by name, got %v", err)
	}
}

// TestAFieldThisContractDoesNotCarryYetIsAccepted is the additive-growth half:
// an engine that adds a field must not turn every already-built rule unknown.
func TestAFieldThisContractDoesNotCarryYetIsAccepted(t *testing.T) {
	grown := strings.Replace(minimal, `"type": "static"`, `"type": "static", "weight": 3`, 1)
	if _, err := ReadEvidence([]byte(grown)); err != nil {
		t.Fatalf("a bundle that gained a field is still a bundle: %v", err)
	}
}

File: packages/archkeep/src/custom-rules/evidence.test.mjs
import { describe, expect, it } from "vitest";

import {
  EVIDENCE_CONTRACT,
  EVIDENCE_KINDS,
  buildEvidenceBundle,
  serializeEvidenceBundle,
} from "./evidence.mjs";

/** One import site in the shape `../analysis/contract.md` fixes. */
const site = {
  sourceFile: "libs/app/main.go",
  line: 7,
  column: 2,
  specifier: "example.test/ring",
  kind: "static",
  spelling: { path: false, relative: false, namesOnly: true },
  resolved: { target: "ring", file: "libs/ring/ring.go", external: false, packageName: null },
};

/** The five observed facts a bundle is built from, all well-formed. */
function observed(overrides = {}) {
  return {
    rule: { name: "no-app-to-ring", reason: "the app layer owns no ring internals" },
    projects: [
      { name: "ring", root: "libs/ring", tags: ["layer-domain"] },
      { name: "app", root: "libs/app", tags: ["layer-app"] },
    ],
    edges: [{ source: "app", target: "ring", type: "static" }],
    imports: [{ site, sourceProject: "app" }],
    policy: {
      depConstraints: [{ sourceTag: "layer-app", onlyDependOnLibsWithTags: ["layer-domain"] }],
      options: { allow: [], banTransitiveDependencies: false },
      suppressions: [],
      fitness: [],
    },
    ...overrides,
  };
}

const decode = (bytes) => new TextDecoder().decode(bytes);

describe("buildEvidenceBundle", () => {
  it("states the contract and all four evidence kinds", () => {
    const bundle = buildEvidenceBundle(observed());
    expect(bundle.contract).toBe(EVIDENCE_CONTRACT);
    // The bundle carries a key per kind, and the roster is the one the host
    // checks a rule's `needs` against — a kind here that is not on the roster
    // (or the reverse) would let a rule declare a need nothing supplies.
    expect(
      Object.keys(bundle)
        .filter((key) => EVIDENCE_KINDS.includes(key))
        .sort(),
    ).toEqual([...EVIDENCE_KINDS].sort());
  });

  it("defaults params in the bundle and leaves the declared row untouched", () => {
    const row = { name: "no-app-to-ring", reason: "why" };
    const bundle = buildEvidenceBundle(observed({ rule: row }));
    expect(bundle.rule).toEqual({ name: "no-app-to-ring", params: {} });
    // The law is the file the workspace committed; the bundle is a view of it.
    expect(Object.hasOwn(row, "params")).toBe(false);
    expect(row).toEqual({ name: "no-app-to-ring", reason: "why" });
  });

  it("carries declared params through and drops everything else on the row", () => {
    const bundle = buildEvidenceBundle(
      observed({ rule: { name: "r", params: { domainTag: "layer-domain" }, reason: "why" } }),
    );
    expect(bundle.rule).toEqual({ name: "r", params: { domainTag: "layer-domain" } });
  });

  it("sorts projects by name without reordering the caller's array", () => {
    const facts = observed();
    const original = [...facts.projects];
    const bundle = buildEvidenceBundle(facts);
    expect(bundle.model.projects.map((project) => project.name)).toEqual(["app", "ring"]);
    expect(facts.projects).toEqual(original);
  });

  it("sorts edges by source, then target, then type, then sourceFile", () => {
    const bundle = buildEvidenceBundle(
      observed({
        edges: [
          { source: "app", target: "ring", type: "static", sourceFile: "libs/app/z.go" },
          { source: "app", target: "ring", type: "static", sourceFile: "libs/app/a.go" },
          { source: "app", target: "ring", type: "dynamic" },
          { source: "app", target: "core", type: "static" },
          { source: "a-first", target: "ring", type: "static" },
        ],
      }),
    );
    expect(
      bundle.graph.edges.map(
        (edge) => `${edge.source}|${edge.target}|${edge.type}|${edge.sourceFile ?? ""}`,
      ),
    ).toEqual([
      "a-first|ring|static|",
      "app|core|static|",
      "app|ring|dynamic|",
      "app|ring|static|libs/app/a.go",
      "app|ring|static|libs/app/z.go",
    ]);
  });

  it("keeps both of two projects a provider named the same", () => {
    // A duplicate name is a provider's bug, and the bundle reporting one
    // project where two were discovered would hide it inside the evidence.
    const bundle = buildEvidenceBundle(
      observed({
        projects: [
          { name: "app", root: "libs/app", tags: [] },
          { name: "app", root: "apps/app", tags: [] },
        ],
      }),
    );
    expect(bundle.model.projects.map((project) => project.root)).toEqual(["libs/app", "apps/app"]);
  });

  it("keeps both of two edges a provider emitted twice", () => {
    // Two rows equal on all four sort keys. Dropping one would be a graph the
    // rule reads as thinner than the one the provider actually reported.
    const twice = { source: "app", target: "ring", type: "static" };
    const bundle = buildEvidenceBundle(observed({ edges: [twice, { ...twice }] }));
    expect(bundle.graph.edges).toEqual([twice, twice]);
  });

  it("keeps an edge's sourceFile off the row when the adapter supplied none", () => {
    const bundle = buildEvidenceBundle(observed());
    expect(Object.hasOwn(bundle.graph.edges[0], "sourceFile")).toBe(false);
  });

  it("passes each import record through verbatim and adds its attribution", () => {
    const bundle = buildEvidenceBundle(observed());
    expect(bundle.imports).toEqual([{ ...site, sourceProject: "app" }]);
    // Verbatim means every key the analysis contract fixes survives, including
    // the ones a graph edge throws away: five of the fifteen boundary rules
    // are decided on `specifier` and `spelling` alone.
    expect(bundle.imports[0].specifier).toBe(site.specifier);
    expect(bundle.imports[0].spelling).toEqual(site.spelling);
  });

  it("keeps imports in the caller's order, which is source order", () => {
    const second = { ...site, line: 2, specifier: "example.test/other" };
    const bundle = buildEvidenceBundle(
      observed({
        imports: [
          { site: { ...site, line: 9 }, sourceProject: "app" },
          { site: second, sourceProject: "app" },
        ],
      }),
    );
    expect(bundle.imports.map((entry) => entry.line)).toEqual([9, 2]);
  });

  it("renames the loaded options to the name the policy file spells", () => {
    const bundle = buildEvidenceBundle(observed());
    expect(bundle.policy.moduleBoundaryOptions).toEqual({
      allow: [],
      banTransitiveDependencies: false,
    });
    expect(bundle.policy.depConstraints).toHaveLength(1);
    // Acceptance is not a rule's business, so what a run waived never reaches
    // one — a rule that could see waivers could launder them into a verdict.
    expect(Object.keys(bundle.policy).sort()).toEqual(["depConstraints", "moduleBoundaryOptions"]);
  });
});

describe("serializeEvidenceBundle", () => {
  it("produces the same bytes for the same facts", () => {
    const first = serializeEvidenceBundle(buildEvidenceBundle(observed()));
    const second = serializeEvidenceBundle(buildEvidenceBundle(observed()));
    expect(first).toEqual(second);
  });

  it("produces the same bytes when a provider hands the same facts in another order", () => {
    const forward = observed();
    const shuffled = observed({
      projects: [...observed().projects].reverse(),
      edges: [
        { source: "app", target: "ring", type: "static" },
        { source: "a-first", target: "ring", type: "static" },
      ],
    });
    shuffled.edges.reverse();
    forward.edges = [
      { source: "a-first", target: "ring", type: "static" },
      { source: "app", target: "ring", type: "static" },
    ];
    expect(decode(serializeEvidenceBundle(buildEvidenceBundle(forward)))).toEqual(
      decode(serializeEvidenceBundle(buildEvidenceBundle(shuffled))),
    );
  });

  it("sorts object keys at every depth, because it is the canonicalizer this package owns", () => {
    const text = decode(serializeEvidenceBundle(buildEvidenceBundle(observed())));
    expect(text.indexOf('"contract"')).toBeLessThan(text.indexOf('"graph"'));
    expect(text.indexOf('"graph"')).toBeLessThan(text.indexOf('"imports"'));
    expect(text.indexOf('"model"')).toBeLessThan(text.indexOf('"policy"'));
    // A nested object is sorted too, which a JSON.stringify of the built
    // object would not do — the two differ only here, so this is the assertion
    // that would go red if the serializer were swapped for a plain stringify.
    const record = JSON.parse(text).imports[0];
    expect(Object.keys(record)).toEqual([...Object.keys(record)].sort());
  });

  it("measures the length in bytes, not in characters", () => {
    // A project name outside the BMP is two UTF-16 code units and four bytes.
    const bundle = buildEvidenceBundle(
      observed({ projects: [{ name: "app-\u{1F600}", root: "libs/app", tags: [] }] }),
    );
    const bytes = serializeEvidenceBundle(bundle);
    expect(bytes).toBeInstanceOf(Uint8Array);
    expect(bytes.byteLength).toBeGreaterThan(decode(bytes).length);
  });
});

describe("what buildEvidenceBundle refuses, by name", () => {
  // Every case here is the silent direction: each of these inputs has an
  // obvious "just carry on" reading — an empty array, a defaulted object —
  // and every one of those readings produces a bundle a rule cannot tell from
  // a workspace that genuinely holds nothing.
  const cases = [
    ["the facts are not an object", () => buildEvidenceBundle(null), "must be an object"],
    ["rule is missing", () => buildEvidenceBundle(observed({ rule: undefined })), "rule:"],
    [
      "rule.name is empty",
      () => buildEvidenceBundle(observed({ rule: { name: " " } })),
      "rule.name",
    ],
    [
      "rule.params is not an object",
      () => buildEvidenceBundle(observed({ rule: { name: "r", params: [] } })),
      "rule.params",
    ],
    [
      "rule.name is a value JSON cannot even render",
      () => buildEvidenceBundle(observed({ rule: { name: 1n } })),
      "bigint 1",
    ],
    [
      "rule.name is a value JSON renders as nothing at all",
      // `JSON.stringify` answers `undefined` for a function rather than
      // throwing, so the refusal would read "rule.name: ... got function
      // undefined" without the fallback to `String`.
      () => buildEvidenceBundle(observed({ rule: { name: () => "no-app-to-ring" } })),
      "function () =>",
    ],
    [
      "the model was never read",
      () => buildEvidenceBundle(observed({ projects: undefined })),
      "the model was never read",
    ],
    [
      "a project is not an object",
      () => buildEvidenceBundle(observed({ projects: [null] })),
      "projects[0] must be an object",
    ],
    [
      "a project has no name",
      () => buildEvidenceBundle(observed({ projects: [{ root: "libs/app", tags: [] }] })),
      "projects[0].name",
    ],
    [
      "a project has no root",
      () => buildEvidenceBundle(observed({ projects: [{ name: "app", tags: [] }] })),
      "projects[0].root",
    ],
    [
      "a project's tags were never read",
      () => buildEvidenceBundle(observed({ projects: [{ name: "app", root: "libs/app" }] })),
      "projects[0].tags",
    ],
    [
      "a project's tags are not strings",
      () =>
        buildEvidenceBundle(observed({ projects: [{ name: "app", root: "libs/app", tags: [7] }] })),
      "projects[0].tags",
    ],
    [
      "the graph was never read",
      () => buildEvidenceBundle(observed({ edges: null })),
      "the graph was never read",
    ],
    [
      "an edge is not an object",
      () => buildEvidenceBundle(observed({ edges: ["app->ring"] })),
      "edges[0] must be an object",
    ],
    [
      "an edge has no target",
      () => buildEvidenceBundle(observed({ edges: [{ source: "app", type: "static" }] })),
      "edges[0].target",
    ],
    [
      "an edge's sourceFile is not a string",
      () =>
        buildEvidenceBundle(
          observed({ edges: [{ source: "app", target: "ring", type: "static", sourceFile: 3 }] }),
        ),
      "edges[0].sourceFile",
    ],
    [
      "nothing was analyzed",
      () => buildEvidenceBundle(observed({ imports: undefined })),
      "nothing was analyzed",
    ],
    [
      "an import entry is not a pair",
      () => buildEvidenceBundle(observed({ imports: [site] })),
      "imports[0].site",
    ],
    [
      "an import entry is not an object at all",
      () => buildEvidenceBundle(observed({ imports: [null] })),
      "imports[0]: must be { site, sourceProject }",
    ],
    [
      "an import site is unattributed",
      () => buildEvidenceBundle(observed({ imports: [{ site, sourceProject: "" }] })),
      "imports[0].sourceProject",
    ],
    ["the policy is missing", () => buildEvidenceBundle(observed({ policy: null })), "policy:"],
    [
      "the policy's constraint rows were never read",
      () => buildEvidenceBundle(observed({ policy: { options: {} } })),
      "policy.depConstraints",
    ],
    [
      "the policy's options were never read",
      () => buildEvidenceBundle(observed({ policy: { depConstraints: [] } })),
      "policy.options",
    ],
  ];

  for (const [what, run, expected] of cases) {
    it(`refuses when ${what}`, () => {
      expect(run).toThrow(/refusing to build an evidence bundle/u);
      expect(run).toThrow(expected);
    });
  }

  it("never lets an absent kind reach a rule as an empty one", () => {
    // The load-bearing half of every case above, stated once: a workspace that
    // genuinely holds no imports serializes `"imports":[]`, so if an ABSENT
    // `imports` also serialized that way the rule could not tell "clean" from
    // "never looked" — and would answer pass for the second.
    const empty = buildEvidenceBundle(observed({ imports: [] }));
    expect(decode(serializeEvidenceBundle(empty))).toContain('"imports":[]');
    expect(() => buildEvidenceBundle(observed({ imports: undefined }))).toThrow();
  });
});
